vm/devices/storage/scsi_buffers/src/lib.rs

```diff
@@ -204,6 +205,52 @@ impl LockedRange for LockedIoVecs {
     }
 }
 
+/// An implementation of [`MemoryWrite`] that provides semantically
+/// correct results. Specifically, it always returns a `ReadOnly` error
+/// when attempting to write to it.
+pub struct PermissionedMemoryWriter<'a> {
+    range: PagedRange<'a>,
+    writer: PagedRangeWriter<'a>,
+    is_write: bool,
+}
+
+impl PermissionedMemoryWriter<'_> {
+    /// Creates a new memory writer with the given range and guest memory.
+    pub fn new<'a>(
+        range: PagedRange<'a>,
+        guest_memory: &'a GuestMemory,
+        is_write: bool,
+    ) -> PermissionedMemoryWriter<'a> {
+        PermissionedMemoryWriter {
+            range,
+            writer: range.writer(guest_memory),
+            is_write,
+        }
+    }
+}
+
+impl MemoryWrite for PermissionedMemoryWriter<'_> {
+    fn write(&mut self, data: &[u8]) -> Result<(), AccessError> {
+        if self.is_write {
+            self.writer.write(data)
+        } else {
+            Err(AccessError::ReadOnly)
+        }
+    }
+
+    fn fill(&mut self, val: u8, len: usize) -> Result<(), AccessError> {
+        if self.is_write {
+            self.writer.fill(val, len)
+        } else {
+            Err(AccessError::ReadOnly)
+        }
+    }
+
+    fn len(&self) -> usize {
+        self.range.len()
```

Comment:
We want this to be correct--it should return 0 for read-only ranges.

But to avoid an extra branch here, I'd suggest having `new` use `PagedRange::empty()` instead of the passed-in range when `!is_write`.

Reply:
Fascinating, I did not immediately grok that `len` returns the space remaining in the underlying thing to be written.